quickwit/quickwit-query/src/query_ast/term_set_query.rs

```diff
@@ -56,15 +63,34 @@ impl TermSetQuery {
                     field: full_path.to_string(),
                     value: value.to_string(),
                 };
-                let ast =
-                    term_query.build_tantivy_ast_call(schema, tokenizer_manager, &[], false)?;
+                let ast = term_query.ast_for_term_extraction(schema, tokenizer_manager)?;
                 let tantivy_query: Box<dyn crate::TantivyQuery> = ast.simplify().into();
+                let mut terms = Vec::new();
                 tantivy_query.query_terms(&mut |term, _| {
-                    terms.insert(term.clone());
+                    terms.push(term.clone());
                 });
+
+                let str_term_count = terms.iter().filter(|term| is_term_str(term)).count();
+                if str_term_count <= 1 {
+                    for term in terms {
+                        all_terms.insert(term);
+                    }
+                } else {
+                    // we have a string, and it got split into multiple tokens, so we want an
+                    // intersection of them
+                    let mut phrase = Vec::with_capacity(terms.len());
+                    for term in terms {
+                        if is_term_str(&term) {
+                            phrase.push(term);
+                        } else {
+                            all_terms.insert(term);
+                        }
+                    }
+                    intersections.push(phrase);
+                }
```

Comment:
The logic is rather odd here. 

You implicitly rely on the fact that you know that the query ast is a disjunction over different field:

e.g.
`(text query OR  numerical query)`

If you want to take that kind of shortcuts (after having considered other solutions), add comments to explain why it works.

Ideally, assert that the query is indeed what you think it is.